### PR TITLE
[Layer] Scaffold layer impl

### PR DESCRIPTION
**Pull request by zhoonit**
- [Layer] Scaffold layer impl

```
**Major changes**
- Create `LayerImpl`
- Change `LayerDevel` signature
  s/int initialize/void finalize
  s/int setProperty/void setProperty
- Move createLayer to layerImpl

**Self evaluation:**
1. Build test: [X]Passed [ ]Failed [ ]Skipped
2. Run test: [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: Jihoon Lee <jhoon.it.lee@samsung.com>

```

### CODE REVIEW

File: nntrainer/layers/layer_impl.h [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+private:
+  bool finalized;                                 /**< check if finalized */
+  std::unique_ptr<std::tuple<>> layer_impl_props; /**< layer_impl_props */
+};
+
+/**
+ * @brief General Layer Factory function to register Layer
+ *
+ * @param props property representation
+ * @return std::unique_ptr<ml::train::Layer> created object
```
Comment (kparichay) [inline nntrainer/layers/layer_impl.h:74]: nntrainer::Layer?

File: nntrainer/layers/layer_impl.h [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+ */
+template <typename T,
```
Comment (kparichay) [inline nntrainer/layers/layer_impl.h:76]: why move to LayerImpl? this will be called by model/graph/etc and they shouldnt have to include layer_impl.

Comment (zhoonit) [reply nntrainer/layers/layer_impl.h:76]: I think the specification is the signature `std::unique_ptr<Layer>(const std::vector<std::string &props>)`. 

If we are going to make this implementation as a specification of an api, works for me.  But I guess we can delay this until we need it (maybe soon 😛 )

Comment (kparichay) [reply nntrainer/layers/layer_impl.h:76]: isn't what the API calls? its left up to you where you keep this but no other header should import layer_impl.h except the ones which derive from LayerImpl (if they do, then a lot other headers will need importing as well). Others should import layer_devel.h

Comment (zhoonit) [reply nntrainer/layers/layer_impl.h:76]: Yes. I'll keep in mind, if this is not possible this must go to layer_devel.h.

Comment (zhoonit) [reply nntrainer/layers/layer_impl.h]: As discussed offline, this is reverted back

File: nntrainer/layers/layer_devel.h [hunk 3]
```diff
@@ -174,6 +171,20 @@ class Layer {
 using CreateLayerFunc = nntrainer::Layer *(*)();
 using DestroyLayerFunc = void (*)(nntrainer::Layer *);
 
+/**
```
Comment (kparichay) [inline nntrainer/layers/layer_devel.h:174]: this is fine for now, but later you can consider putting `LayerPluggable` stuff together, and moving `createLayer` out from the middle of it's related declarations and definitions.

**Review by kparichay — APPROVED** [review]
LGTM

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_impl.h [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+class Exporter;
+
+enum class ExportMethods;
+/**
+ * @class   An abstract class to ease developing a layer
+ * @brief   An abstract class for all layers
+ *
+ */
+class LayerImpl : public Layer {
+
```
Comment (jijoongmoon) [inline nntrainer/layers/layer_impl.h:35]: This ```public Layer``` is from layer_devel?

**Review by jijoongmoon — APPROVED** [review]
LGTM

File: nntrainer/layers/layer_impl.h (hunk 4 above)
Comment (zhoonit) [reply nntrainer/layers/layer_impl.h:35]: Yes, it is from Layer :)

**Pull request merged**